### ISSUE TITLE
Warnings or auto-draining about not touched HttpRequest dataBytes

### ISSUE THREAD

**Issue by ktoso** [akka/akka-http#183]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [ktoso](https://github.com/ktoso)**
_Tuesday Oct 13, 2015 at 08:22 GMT_
_Originally opened as https://github.com/akka/akka/issues/18716_

---

In the bypass merge we can figure out that once the user has applied the Request => Response,
we require the dataBytes source to have been materialized.
If it's not, we log a warning, and we could drain the dataBytes.

Split when will need to emit a special source, that we can inspect and see if it's been materialized or not.

**Comment by ktoso** [comment]
<a href="https://github.com/drewhk"><img src="https://avatars.githubusercontent.com/u/932285?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [drewhk](https://github.com/drewhk)**
_Tuesday Oct 13, 2015 at 08:23 GMT_

----

warnings *and* auto-draining.

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Tuesday Oct 13, 2015 at 08:23 GMT_

----

We're working on fusing and things around Http, and this idea came up.
We'll be able to warn or consume the dataBytes for the users if the user forgot to drain the data from the request.

*"Single actor Http pipeline, here we come!"*

// cc @sirthias @jrudolph

**Comment by ktoso** [comment]
<a href="https://github.com/jrudolph"><img src="https://avatars.githubusercontent.com/u/9868?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jrudolph](https://github.com/jrudolph)**
_Tuesday Oct 13, 2015 at 09:08 GMT_

----

You may want to look at existing tickets. There are some that suggest similar things but don't mention the technical approach.

> "Single actor Http pipeline, here we come!"

So, `concatAll` will be fusable?

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Tuesday Oct 13, 2015 at 09:30 GMT_

----

Both concat and splitWhen will be mergable (i.e. in general streams of streams will be)

**Comment by ktoso** [comment]
<a href="https://github.com/drewhk"><img src="https://avatars.githubusercontent.com/u/932285?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [drewhk](https://github.com/drewhk)**
_Tuesday Oct 13, 2015 at 09:34 GMT_

----

this ticket actually has nothing to do with fusing though, it just happened that we discussed this during our fusion discussion.

**Comment by ktoso** [comment]
<a href="https://github.com/sirthias"><img src="https://avatars.githubusercontent.com/u/135347?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sirthias](https://github.com/sirthias)**
_Tuesday Oct 13, 2015 at 12:45 GMT_

----

Interesting.
Note that auto-draining might not always be what I want, though!
For example, if the request contains an entity with a content-length that it beyond the configured threshold I want to send an error response and *NOT* drain the incoming byte stream.
So, this auto-draining would definitely have to have a config switch (which could default to *on*).

**Comment by ktoso** [comment]
<a href="https://github.com/drewhk"><img src="https://avatars.githubusercontent.com/u/932285?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [drewhk](https://github.com/drewhk)**
_Tuesday Oct 13, 2015 at 13:21 GMT_

----

yes, that makes sense. We also want to log a Warning though, so it is not as much *on* vs *off, but instead *warn* vs *fail*?

**Comment by ktoso** [comment]
<a href="https://github.com/sirthias"><img src="https://avatars.githubusercontent.com/u/135347?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sirthias](https://github.com/sirthias)**
_Tuesday Oct 13, 2015 at 13:51 GMT_

----

Wouldn't it be something like

```
auto-drain-request-entity = on-with-warning # or `off`
```

?

**Comment by ktoso** [comment]
<a href="https://github.com/drewhk"><img src="https://avatars.githubusercontent.com/u/932285?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [drewhk](https://github.com/drewhk)**
_Tuesday Oct 13, 2015 at 13:56 GMT_

----

maybe that. I am not sure, I think we will see how it works in practice in the end anyway. This is just a high level idea so far.

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Wednesday Dec 16, 2015 at 10:56 GMT_

----

For reference, bypassMerge is gone, but the idea of this ticket is still valid.

**Comment by ktoso** [comment]
<a href="https://github.com/rkuhn"><img src="https://avatars.githubusercontent.com/u/470469?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [rkuhn](https://github.com/rkuhn)**
_Monday Mar 21, 2016 at 13:46 GMT_

----

In addition to the discussed possibilities it could make sense to add an option that auto-drains the request entity for successful “early” responses, but not for unsuccessful ones—in the latter case it would emit the error and close the connection (because continuing afterwards would not be legal for HTTP). This yields the following list of possibilities:

* no detection (current behavior)
* warning for unconsumed entities
* auto-drain unless rejected, in which case early response and close (should be default)
* auto-drain always

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Thursday May 19, 2016 at 00:10 GMT_

----

Relates to https://github.com/akka/akka/issues/19602

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Saturday Jul 30, 2016 at 17:50 GMT_

----

ping @rkuhn, if you have any new ideas please share :-)
We have not been working on this so far btw.

**Comment by toebbel** [comment]
I am interested in seeing similar warnings for `ResponseEntities`. We had a case, where we removed some lines that would consume a http `ResponseEntity` from an upstream service, however, the request was still being fired. Since a cached host http connection pool is used, the allocated connections were never free'ed since their responses were neither read nor discarded.

Coming from `http4s` and `fs2.Task` we did not expect that `Http().singleRequest` performs the side-effect of checking out a connection of the pool, before the response entity is touched.
